subsys/nsettings/settings_fcb.c

```diff
@@ -0,0 +1,550 @@
+/* Copyright (c) 2019 Laczen
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ * Copyright (c) 2015 Runtime Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <string.h>
+
+#include "settings_fcb.h"
+#include "settings_priv.h"
+
+struct settings_fcb_read_fn_arg {
+	struct fcb_entry_ctx *entry_ctx;
+	off_t off;
+};
+
+static int settings_fcb_load(struct settings_store *cs);
+static int settings_fcb_save(struct settings_store *cs, const char *name,
+			     const char *value, size_t val_len);
+
+static struct settings_store_itf settings_fcb_itf = {
+	.csi_load = settings_fcb_load,
+	.csi_save = settings_fcb_save,
+};
+
+int settings_fcb_dst(struct settings_fcb *cf)
+{
+	cf->cf_store.cs_itf = &settings_fcb_itf;
+	settings_dst_register(&cf->cf_store);
+
+	return 0;
+}
+
+int settings_fcb_src(struct settings_fcb *cf)
+{
+	cf->cf_store.cs_itf = &settings_fcb_itf;
+	settings_src_register(&cf->cf_store);
+
+	return 0;
+}
+
+static int settings_fcb_read(struct fcb_entry_ctx *entry_ctx, off_t off,
+			     char *buf, size_t *len)
+{
+	if (off >= entry_ctx->loc.fe_data_len) {
+		return -EINVAL;
+	}
+
+	if ((off + *len) > entry_ctx->loc.fe_data_len) {
+		*len = entry_ctx->loc.fe_data_len - off;
+	}
+
+	return flash_area_read(entry_ctx->fap,
+			       FCB_ENTRY_FA_DATA_OFF(entry_ctx->loc) + off,
+			       buf, *len);
+}
+
+static ssize_t settings_fcb_read_fn(void *data, size_t len, void *read_fn_arg)
+{
+	int rc, len_read;
+	struct settings_fcb_read_fn_arg *rd_fn_arg;
+
+	len_read = len;
+	rd_fn_arg = (struct settings_fcb_read_fn_arg *)read_fn_arg;
+	rc = settings_fcb_read(rd_fn_arg->entry_ctx, rd_fn_arg->off,
+			       data, &len_read);
+	if (rc < 0) {
+		return rc;
+	}
+	return len_read;
+}
+
+static int settings_fcb_load(struct settings_store *cs)
```

Comment:
I think resignation from settings_line_xxx routines is not necessary. This was common for fcb and FS. I see that bas64 was eliminated as well (I' agree with this). It shouldn't be so hard to keep these common even while beck-end coexistence.

Reply:
I'm open for counter arguments.

Reply:
You are right, but here is my reasoning: in fcb the data is: name=bin(value), in file it is: name=len(value)bin(value). As they are no longer the same (they were when encoded) it makes no sense to try and create some intermediate line format. This also makes it easier to support multiple backends at the same time. I think it is also easier to understand the code and maintain.

Reply:
I can understand that operation on FS back-end is so simple that no sense to implement it common with fcb. but line assembly and parsing (to RAW flash) is not only-specific for fcb. In feature it might be required for another storage module which doesn't exist  now (with very probably will come, just by probability). I'm thinking about it as about blocks for building the a back-end. So even if line methods are only for fcb for now - it right be fruitful to keep them extracted.
Additionally it always helps to review stuff.

Reply:
I agree, but I think what will mainly change when another backend is added is exactly the way data is stored. Think about storage in json format, would the line be a good format choice ?

Reply:
I'm thinking about keep some routines as redy to be used elements - not nesesery enforce it usage for other back-end.

Reply:
In my point of view, removing settings_line_xxx made the subsystem easier to understand. While maintaining common building blocks is a good idea, it might become harder to maintain if "line-like" backends start adding support for their own specific variants to a common settings_line.c (like the current USE_BASE64 and ENCODE_LEN KConfig options).